driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -289,7 +291,13 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
             // At startup, when we add the initial nodes, this will be null, which is ok
             if (c == null)
                 return;
-            refreshSchema(c, keyspace, table, cluster, false);
+            refreshSchema(c, keyspace, table, cluster);
+            // If the table is null, we either rebuild all from scratch or have an updated keyspace. In both cases, rebuild the token map
+            // since some replication on some keyspace may have changed
+            if (table == null) {
+                cluster.submitNodeListRefresh();
+                cluster.nodeListRefreshRequestDebouncer.deliverEvents().get();
```

Comment:
I don't think we need to wait on the debouncer here: if the token map is stale for a while that's not a big deal, so we don't need to wait synchronously.
I would simply submit the refresh and carry on. I'm working on some changes, I'll include that.